src/native/fs/fs_napi.cpp

```diff
@@ -0,0 +1,747 @@
+        //   InstanceMethod("write", &FileWrap::write),
+        });
+        constructor = Napi::Persistent(func);
+        constructor.SuppressDestruct();
+    }
+    FileWrap(const Napi::CallbackInfo& info)
+        : Napi::ObjectWrap<FileWrap>(info)
+        , _fd(0)
+    {
+    }
+    ~FileWrap() {
+        if (_fd) {
+            PANIC("FS::FileWrap::dtor: file not closed " << DVAL(_path) << DVAL(_fd));
+        }
+    }
+    Napi::Value close(const Napi::CallbackInfo& info);
+    Napi::Value read(const Napi::CallbackInfo& info);
+    // Napi::Value write(const Napi::CallbackInfo& info);
+};
+
+Napi::FunctionReference FileWrap::constructor;
+
+struct FileOpen : public FSWorker
+{
+    std::string _path;
+    int _fd;
+    FileOpen(const Napi::CallbackInfo& info) : FSWorker(info), _fd(0)
+    {
+        _path = info[0].As<Napi::String>();
+        // TODO - info[1] { mode, readonly }
+        Begin(XSTR() << DVAL(_path));
+    }
+    virtual void Execute()
+    {
+        _fd = open(_path.c_str(), O_RDONLY); // TODO mode
+        if (!_fd) SetSyscallError();
+    }
+    virtual void OnOK()
+    {
+        LOG("FS::DirOpen::OnOK: " << DVAL(_path));
+        Napi::Object res = FileWrap::constructor.New({});
+        FileWrap *w = FileWrap::Unwrap(res);
+        w->_path = _path;
+        w->_fd = _fd;
+        _deferred.Resolve(res);
+    }
+};
+
+struct FileClose : public FSWorker
+{
+    FileWrap *_wrap;
+    FileClose(const Napi::CallbackInfo& info) : FSWorker(info)
+    {
+        _wrap = FileWrap::Unwrap(info.This().As<Napi::Object>());
+    }
+    virtual void Execute()
+    {
+        int fd = _wrap->_fd;
+        std::string path = _wrap->_path;
+        int r = close(fd);
+        if (r) SetSyscallError();
+        _wrap->_fd = 0;
+    }
+};
+
+
+struct FileRead : public FSWorker
+{
+    FileWrap *_wrap;
+    uint8_t* _buf;
+    off_t _offset;
+    int _len;
+    int _pos;
+    ssize_t _br;
+    FileRead(const Napi::CallbackInfo& info) 
+        : FSWorker(info) 
+        , _buf(0)
+        , _offset(0)
+        , _len(0)
+        , _pos(0)
+    {
+        _wrap = FileWrap::Unwrap(info.This().As<Napi::Object>());
+        auto buf = info[0].As<Napi::Buffer<uint8_t>>();
+        _buf = buf.Data();
+        _offset = info[1].As<Napi::Number>();
+        _len = info[2].As<Napi::Number>();
+        _pos = info[3].As<Napi::Number>();
+    }
+    virtual void Execute()
+    {
+        int fd = _wrap->_fd;
+        std::string path = _wrap->_path;
+        if (fd < 0) {
+            SetError(XSTR() << "FS::FileRead::Execute: ERROR not opened " << path);
+            return;
+        }
+        _br = pread(fd, _buf + _offset, _len, _pos);
+        if (_br < 0) {
+            SetSyscallError();
+            return;
+        }
+    }
+    virtual void OnOK()
+    {
+        Napi::Env env = Env();
+        _deferred.Resolve(Napi::Number::New(env, _br));
+    }
+};
+
+// struct FileWrite : public FSWorker
```

Comment:
Are you going to complete this op in this PR?

Reply:
Next PR, I will also probably rewrite the logic for uploads and reads to use the FileWrapper